horovod/mxnet/mpi_ops.cc

```diff
@@ -20,60 +20,43 @@
 #include "../common/operations.h"
 #include "adapter.h"
 #include "cuda_util.h"
-#include "handle_manager.h"
 #include "mpi_ops.h"
 #include "ready_event.h"
 #include "tensor_util.h"
 
 namespace horovod {
 namespace mxnet {
 
-static HandleManager handle_manager;
-
 namespace {
-
 std::string GetOpName(std::string prefix, char* name) {
   if (name != nullptr) {
     return prefix + "." + std::string(name);
   }
   return std::string();
 }
-
-std::string GetOpNameHandle(std::string prefix, const std::string& name,
-                            int handle) {
-  if (name.length() != 0) {
-    return name;
-  }
-  return prefix + ".noname." + std::to_string(handle);
-}
-
 } // namespace
 
-int DoAllreduce(NDArray* tensor, NDArray* output, const std::string& name,
-                Callback cb) {
+void DoAllreduce(NDArray* tensor, NDArray* output, const std::string& name,
+                 Callback cb) {
   ThrowIfError(common::CheckInitialized());
 
-  auto handle = handle_manager.AllocateHandle(cb);
   auto device = TensorUtil::GetDevice(tensor);
   auto hvd_tensor = std::make_shared<MXTensor<NDArray>>(tensor);
   auto hvd_context = std::make_shared<MXOpContext<NDArray>>(device, output);
   auto hvd_output = std::make_shared<MXTensor<NDArray>>(output);
 
   auto enqueue_result =
       EnqueueTensorAllreduce(hvd_context, hvd_tensor, hvd_output, nullptr,
-                             GetOpNameHandle("allreduce", name, handle), device,
```

Comment:
@ctcyang could you please elaborate on the 1st point? I am looking at the commit you provided. The commit tried to replace char* name with const std::string& name in DoXXX functions. Without this change, if const* name is nullptr, you will generate a string prefix + ".noname" + std::to_string(handle) to pass to EnqueueTensorXXX functions. With this change, if const* name is nullptr, we will pass an empty string. And this empty string will cause problem?

Reply:
Yeah, my concern is that if we pass an empty string, there will non-deterministically be multiple NDArrays that will be associated with an empty string, which will cause problems down the line when Horovod tries to do allreduce and broadcast on NDArrays of different sizes.

Reply:
It is valid point. @apeforest can we return a non-empty string in GetOpName? Instead of returning std::string() while name == nullptr, we can return something like prefix + ".noname". @ctcyang does the name needs to be unique? If yes, we can append a random number or the current time at the end.

Reply:
Yeah, I believe the name needs to be unique. The way it was done before was by incrementing by one in `handle_manager`: https://github.com/ctcyang/horovod/blob/0128af0a726348b708f246a3643a0dfde09edf51/horovod/mxnet/handle_manager.cc#L24 I am surprised this number was consistent across MPI processes.

Reply:
@ctcyang I still don't understand why this would cause a problem. The allreduce, allgather, broadcast APIs are called with a non-empty name string passed in: https://github.com/ctcyang/horovod/blob/develop/mxnet/horovod/mxnet/__init__.py

I don't find any problem running locally on my Mac. I am verifying on p3x16 at the moment.

Reply:
Updated the GetOpName per your suggestion. Please review again.

Reply:
Did you run into any issues with p3x16 with the previous commit that were fixed with the new `GetOpName`?

Reply:
Regarding the second point raised by @ctcyang, I think we should be good after revisiting the logic. The previous nullptr check was needed because we store the callback in a map which has concurrent rw. We will be running performance benchmarks and will be able to validate the stability after removing handle manager.